script/src/verify.rs

```diff
@@ -1117,21 +1133,26 @@ fn run_vms(
     };
 
     while !machines.is_empty() {
-        let (mut machine, resume_data) = machines.pop().unwrap();
+        let mut machine = machines.pop().unwrap();
 
-        if let Some((callee_exit_code, callee_cycles, callee_resume_data)) = callee_data {
+        if let (Some((callee_exit_code, callee_cycles)), Some(callee_spawn_data)) =
+            (callee_data, &spawn_data)
+        {
             update_caller_machine(
-                &mut machine.machine,
+                &mut machine.machine_mut().machine,
                 callee_exit_code,
                 callee_cycles,
-                &callee_resume_data,
+                callee_spawn_data,
             )
             .map_err(map_vm_internal_error)?;
         }
 
         match machine.run() {
             Ok(code) => {
-                callee_data = Some((code, machine.machine.cycles(), resume_data));
+                callee_data = Some((code, machine.cycles()));
+                if let ResumableMachine::Spawn(_, data) = machine {
+                    spawn_data = Some(data);
+                }
```

Comment:
```
if let ResumableMachine::Spawn(_, spawn_data) = machine {
  callee_data = Some((code, machine.cycles(), spawn_data));
}
```

Reply:
This won't work, consider a case where we have a single script that does not call spawn, it gets suspended somehow. In case case, we will need `code` and `machine.cycles()` to update the final return result, but `spawn_data` will be missing.

Reply:
But I see that on line 1138 we did match `(Some, Some)`, in the case you describe, is there a problem too? Can you add a corresponding unit test?

Reply:
Imagine a case where we have 3 scripts:

```
A -> B -> C
```

A calls B via spawn, B calls C via spawn. And during the execution of C, the VM gets suspended.

The code works like following on line 1183:

* C is the leaf VM, line 1138 will result in a miss. We don't need to update any information on C(which makes sense, C is the leaf VM, it is not any caller VM).
* When C terminates, `callee_data` is updated with exit code and cycles from C, `spawn_data` contains any data C set via `set_content`
* At this stage we will continue run B, line 1138 will result in a match since both `callee_data` and `spawn_data` are available, which makes sense, since B indeed calls spawn, it would require the exit code of a child script, and any data the child script updates via `set_content`. The cycles consumed to run C will also be added to B.
* When B terminates, it updates `callee_data` and `spawn_data` again.
* At this stage we will run A, line 1138 will result in a match. `callee_data` and `spawn_data` will be used to update A to reflect spawn results.
* When A terminates, it updates `callee_data`, but since it is not created from a spawn syscall, it does not update `spawn_data` per line 1153.
* At the final stage, we do `unwrap` on `callee_data`, to get exit code and cycles for the complete execution trace.

Now consider a different case:

```
A
```

We only executes A, A does not call spawn, the VM gets suspended when running A:

* Upon entry, line 1138 will be a miss since `callee_data` and `spawn_data` are both `None`.
* When A completes execution, it will update `callee_data` on line 1152, but it is not created from a spawn syscall, it does not update `spawn_data` per line 1153.
* Finally, we do `unwrap` on `callee_data`, to get the exit code and cycles for the complete execution trace.

In this second case, we will run into a scenario where `callee_data` contains data, but `spawn_data` stays None.

There is never a case when `spawn_data` contains data but `callee_data` is None. `callee_data` always contains the exit code and cycles for the last terminated machine.

Actually all the above behavior has been covered by tests already.

Reply:
I see, I didn't notice at first that there was an unwrap after the loop, thanks for the detailed explanation.

but I think we can use a simpler variable initialization here to remove the `Option` init / unwrap:

```rust
let (mut exit_code, mut cycles, mut spawn_data) = (0, 0, None);

while let Some(mut machine) = machines.pop() {
    if let Some(callee_spawn_data) = spawn_data.as_ref() {
        update_caller_machine(
...

    match machine.run() {
        Ok(code) => {
            exit_code = code;
            cycles = machine.cycles();
            if let ResumableMachine::Spawn(_, data) = machine {
                spawn_data = Some(data);
            }
```

Reply:
Done